### PR TITLE
Device can now remove itself

### PR DESCRIPTION
**Pull request by kosssi**
It's possible to create new release after merge it ?

### CODE REVIEW

**Comment by nono** [comment]
@kosssi you should make the PR on the `development` branch, and without building JS files

File: server/controllers/devices.coffee [hunk 1]
```diff
@@ -93,7 +93,9 @@ initAuth = (req, cb) ->
     [username, password] = extractCredentials req.headers['authorization']
     # Initialize user
     user = {}
-    user.body = password: password
+    user.body =
+        username: username
+        password: password
```
Comment (nono) [inline server/controllers/devices.coffee]: `initAuth` is used by several functions. Are you sure your change has no side effect on them?

File: server/controllers/devices.coffee [hunk 2]
```diff
@@ -241,30 +243,42 @@ module.exports.update = (req, res, next) ->
...
-
-            checkLogin login, true, (err, device) ->
-                return next err if err?
-                # Remove device
-                removeDevice device, (err) ->
-                    if err?
-                        next err
-                    else
-                        res.sendStatus 204
+    callback = (next) ->
```
Comment (nono) [inline server/controllers/devices.coffee]: You could find a name more descriptive than `callback` here

File: server/controllers/devices.coffee [hunk 3]
```diff
@@ -241,30 +243,42 @@ module.exports.update = (req, res, next) ->
...
+                if auth
+                    callback next
+                else
+                    error = new Error "Request unauthorized"
+                    error.status = 401
+                    next error
+        else
+            passport.authenticate 'local', (err, user) =>
+                if err
+                    console.log err
```
Comment (nono) [inline server/controllers/devices.coffee]: `console.log` spotted

File: server/controllers/devices.coffee [hunk 4]
```diff
@@ -241,30 +243,42 @@ module.exports.update = (req, res, next) ->
...
     initAuth req, (user) ->
-        # Check if request is authenticated
-        authenticator user, res
+        username = user.body.username
+        password = user.body.password
+
+        # Test if login is equal to the deviceName
+        # a device can remove itself.
+        if username is req.params.login
+            deviceManager.isAuthenticated username, password, (auth) =>
```
Comment (nono) [inline server/controllers/devices.coffee]: I don't think the fat arrow is needed here

File: test/devices_test.coffee [hunk 5]
```diff
@@ -183,10 +185,17 @@ describe "Devices", ->
 
         describe 'Delete a device', ->
 
-            it "Delete a device", (done) ->
+            it "can't delete a device from another device login", (done) ->
+                @timeout 10 * 1000
+                client.setBasicAuth  'test_device_2', @password
+                client.del "device/test_device", (err, res, body) =>
+                    should.exist err
```
Comment (nono) [inline test/devices_test.coffee]: I'd have test the HTTP response code here (403, I suppose)

**Comment by kosssi** [comment]
the build is red. I test some refactoring.

**Comment by frankrousseau** [comment]
@kosssi What's the status of this PR?

**Comment by frankrousseau** [comment]
ping @kosssi

**Comment by kosssi** [comment]
I finish this PR today :eyeglasses:

**Comment by kosssi** [comment]
I change to development branch : https://github.com/cozy/cozy-proxy/pull/263

I don't forget yours comments ;)

**Pull request closed without merging**